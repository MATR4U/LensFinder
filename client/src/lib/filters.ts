import type { Lens } from '../types';
import { coverageMatches } from './availability';

export type FiltersInput = {
  lenses: Lens[];
  cameraName: string;
  cameraMount?: string;
  brand: string;
  lensType: string;
  sealed: boolean;
  isMacro: boolean;
  priceRange: { min: number; max: number };
  weightRange: { min: number; max: number };
  proCoverage: string;
  proFocalMin: number;
  proFocalMax: number;
  proMaxApertureF: number;
  proRequireOIS: boolean;
  proRequireSealed: boolean;
  proRequireMacro: boolean;
  proPriceMax: number;
  proWeightMax: number;
  proDistortionMaxPct: number;
  proBreathingMinScore: number;
  // Soft vs hard toggles
  softPrice?: boolean;
  softWeight?: boolean;
  softDistortion?: boolean;
  softBreathing?: boolean;
};

export function applyFilters(input: FiltersInput): Lens[] {
  const {
    lenses,
    cameraName,
    cameraMount,
    brand,
    lensType,
    sealed,
    isMacro,
    priceRange,
    weightRange,
    proCoverage,
    proFocalMin,
    proFocalMax,
    proMaxApertureF,
    proRequireOIS,
    proRequireSealed,
    proRequireMacro,
    proPriceMax,
    proWeightMax,
    proDistortionMaxPct,
    proBreathingMinScore
  } = input;

  const base = cameraMount && cameraName !== 'Any' ? lenses.filter(l => l.mount === cameraMount) : lenses.slice();
  const applyFocalRange = (proFocalMin > 0) || (proFocalMax < 9999);

  return base
    .filter((l) => (brand === 'Any' ? true : l.brand === brand))
    .filter((l) => {
      const type = l.focal_min_mm === l.focal_max_mm ? 'Prime' : 'Zoom';
      return lensType === 'Any' ? true : type === lensType;
    })
    .filter((l) => (sealed ? l.weather_sealed : true))
    .filter((l) => (isMacro ? l.is_macro : true))
    .filter((l) => (input.softPrice ? true : (l.price_chf >= priceRange.min && l.price_chf <= priceRange.max)))
    .filter((l) => (input.softWeight ? true : (l.weight_g >= weightRange.min && l.weight_g <= weightRange.max)))
    .filter((l) => coverageMatches(l.coverage, proCoverage))
    .filter((l) => !applyFocalRange ? true : ((l.focal_min_mm ?? 0) <= proFocalMin && (l.focal_max_mm ?? 0) >= proFocalMax))
    .filter((l) => (l.aperture_min ?? 99) <= proMaxApertureF)
    .filter((l) => (proRequireOIS ? !!l.ois : true))
    .filter((l) => (proRequireSealed ? !!l.weather_sealed : true))
    .filter((l) => (proRequireMacro ? !!l.is_macro : true))
    .filter((l) => (l.price_chf ?? Infinity) <= proPriceMax)
    .filter((l) => (l.weight_g ?? Infinity) <= proWeightMax)
    .filter((l) => (input.softDistortion ? true : ((l.distortion_pct ?? 0) <= proDistortionMaxPct)))
    .filter((l) => (input.softBreathing ? true : ((l.focus_breathing_score ?? 0) >= proBreathingMinScore)));
}


