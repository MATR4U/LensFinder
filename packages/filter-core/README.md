Filter Core

Exports
- Types: FilterSpec, RankOptions, DatasetSchema, Clause, Executable, Explanation
- API: compile, filter, rank, explain
- Adapters: fromStore, fromQueryParams, toQueryParams
- OpenAPI helpers: fetchPaged, schemaToDataset, paramsToSpec

Usage
import { compile, filter, rank } from '@lensfinder/filter-core';

const spec = { allOf: [{ path: 'price', op: 'lte', value: 1000, mode: 'hard' }, { path: 'weight', op: 'lte', value: 800, mode: 'soft', weight: 1 }] };
const exec = compile(spec);
const filtered = filter(items, exec);
const ranked = rank(items, exec, { topK: 10 });
