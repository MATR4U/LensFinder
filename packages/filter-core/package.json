{
  "name": "@lensfinder/filter-core",
  "version": "0.1.0",
  "private": true,
  "type": "module",
  "main": "dist/cjs/index.cjs",
  "module": "dist/esm/index.js",
  "types": "dist/types/index.d.ts",
  "exports": {
    ".": {
      "types": "./dist/types/index.d.ts",
      "import": "./dist/esm/index.js",
      "require": "./dist/cjs/index.cjs"
    }
  },
  "sideEffects": false,
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "npm run clean && npm run build:types && npm run build:esm && npm run build:cjs",
    "build:esm": "tsc -p tsconfig.build.json --module ESNext --outDir dist/esm",
    "build:cjs": "tsc -p tsconfig.build.cjs.json && node scripts/cjs-post.cjs",
    "build:types": "tsc -p tsconfig.types.json",
    "clean": "rimraf dist",
    "test": "vitest run"
  },
  "devDependencies": {
    "@types/node": "^20.12.12",
    "ajv": "^8.17.1",
    "fast-check": "^3.18.0",
    "rimraf": "^5.0.5",
    "typescript": "^5.4.5",
    "vitest": "^1.6.0"
  }
}
