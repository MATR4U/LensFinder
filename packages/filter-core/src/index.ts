export * from './types';
export { compile } from './compile/compiler';
export { filter } from './engine/filter';
export { rank } from './engine/rank';
export { explain } from './engine/explain';
export { fetchPaged, schemaToDataset, paramsToSpec } from './providers/httpOpenApi';
export { fromStore } from './adapters/zustand';
export { fromQueryParams, toQueryParams } from './adapters/router';
export { buildSpecFromState, applyWithPrefilters } from './adapters/app';
export type { Mapping, PreFilter } from './adapters/app';
export { toURLSearchParams } from './adapters/next';
export { validateSpec } from './openapi/validate';
export { HNSWIndex } from './ann/hnsw';
