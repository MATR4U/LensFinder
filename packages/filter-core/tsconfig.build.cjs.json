{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "composite": false,
    "declaration": false,
    "noEmit": false,
    "module": "CommonJS",
    "moduleResolution": "Node",
    "outDir": "dist/cjs"
  },
  "include": ["src"]
}
