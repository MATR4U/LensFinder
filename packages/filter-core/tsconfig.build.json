{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "composite": false,
    "declaration": false,
    "noEmit": false,
    "module": "ESNext"
  },
  "include": ["src"]
}
