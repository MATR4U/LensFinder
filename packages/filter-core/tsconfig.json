{
  "compilerOptions": {
    "composite": true,
    "rootDir": "src",
    "skipLibCheck": true,
    "resolveJsonModule": true,
    "target": "ES2020",
    "module": "ESNext",
    "moduleResolution": "Bundler",
    "lib": ["ES2020", "DOM"],
    "strict": true,
    "noEmit": true,
    "isolatedModules": true
  },
  "include": ["src"]
}
