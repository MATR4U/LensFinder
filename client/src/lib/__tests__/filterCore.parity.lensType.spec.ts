import { describe, it, expect } from 'vitest';
import { applyFilters, type FiltersInput } from '../filters';
import { compile, filter as fcFilter, type FilterSpec, buildSpecFromState, type Mapping } from '@lensfinder/filter-core';
import type { Lens } from '../../types';
import lensesFixture from '../../../../server/tests/fixtures/lenses.json';
const allLenses = lensesFixture as unknown as Lens[];

function legacy(input: FiltersInput): Lens[] {
  return applyFilters(input);
}

function core(input: FiltersInput): Lens[] {
  const base = (input.cameraMount && input.cameraName !== 'Any')
    ? input.lenses.filter(l => l.mount === input.cameraMount)
    : input.lenses.slice();
  const preLensType = base.filter(l => {
    const type = l.focal_min_mm === l.focal_max_mm ? 'Prime' : 'Zoom';
    return input.lensType === 'Any' || !input.lensType ? true : type === input.lensType;
  });
  const mapping: Mapping = [
    { from: 'brand', to: 'brand', op: 'eq', when: (s: any) => s.brand && s.brand !== 'Any' },
    { from: 'sealed', to: 'weather_sealed', op: 'isTrue', when: (s: any) => s.sealed === true },
    { from: 'isMacro', to: 'is_macro', op: 'isTrue', when: (s: any) => s.isMacro === true },
    { from: 'priceRange', to: 'price_chf', op: 'between', mode: (st: any) => (st.softPrice ? 'soft' : 'hard'), transform: (v: any) => [ v.min, v.max ], weight: (st: any) => (st.softPrice ? 1 : undefined) },
    { from: 'weightRange', to: 'weight_g', op: 'between', mode: (st: any) => (st.softWeight ? 'soft' : 'hard'), transform: (v: any) => [ v.min, v.max ], weight: (st: any) => (st.softWeight ? 1 : undefined) },
    { from: 'proMaxApertureF', to: 'aperture_min', op: 'lte' },
    { from: 'proPriceMax', to: 'price_chf', op: 'lte' },
    { from: 'proWeightMax', to: 'weight_g', op: 'lte' },
    { from: 'proRequireOIS', to: 'ois', op: 'isTrue', when: (s: any) => s.proRequireOIS === true },
    { from: 'proRequireSealed', to: 'weather_sealed', op: 'isTrue', when: (s: any) => s.proRequireSealed === true },
    { from: 'proRequireMacro', to: 'is_macro', op: 'isTrue', when: (s: any) => s.proRequireMacro === true },
    { from: 'proDistortionMaxPct', to: 'distortion_pct', op: 'lte', mode: (st: any) => (st.softDistortion ? 'soft' : 'hard'), weight: (st: any) => (st.softDistortion ? 1 : undefined) },
    { from: 'proBreathingMinScore', to: 'focus_breathing_score', op: 'gte', mode: (st: any) => (st.softBreathing ? 'soft' : 'hard'), weight: (st: any) => (st.softBreathing ? 1 : undefined) },
  ];
  const spec = buildSpecFromState({ ...input }, mapping) as FilterSpec;
  const exec = compile(spec as FilterSpec);
  return fcFilter(preLensType, exec) as Lens[];
}

function makeInput(overrides: Partial<FiltersInput> = {}): FiltersInput {
  return {
    lenses: allLenses,
    cameraName: 'Any',
    cameraMount: undefined,
    brand: 'Any',
    lensType: 'Any',
    sealed: false,
    isMacro: false,
    priceRange: { min: 0, max: 100000 },
    weightRange: { min: 0, max: 10000 },
    proCoverage: 'Any',
    proFocalMin: 0,
    proFocalMax: 10000,
    proMaxApertureF: 32,
    proRequireOIS: false,
    proRequireSealed: false,
    proRequireMacro: false,
    proPriceMax: 100000,
    proWeightMax: 10000,
    proDistortionMaxPct: 100,
    proBreathingMinScore: 0,
    softPrice: false,
    softWeight: false,
    softDistortion: false,
    softBreathing: false,
    enablePrice: true,
    enableWeight: true,
    enableDistortion: true,
    enableBreathing: true,
    ...overrides,
  };
}

describe('parity: lens type gating', () => {
  it('Any', () => {
    const input = makeInput({ lensType: 'Any' });
    const a = legacy(input);
    const b = core(input);
    expect(new Set(b.map(x => x.source_url || x.name))).toEqual(new Set(a.map(x => x.source_url || x.name)));
  });
  it('Prime', () => {
    const input = makeInput({ lensType: 'Prime' });
    const a = legacy(input);
    const b = core(input);
    expect(new Set(b.map(x => x.source_url || x.name))).toEqual(new Set(a.map(x => x.source_url || x.name)));
  });
  it('Zoom', () => {
    const input = makeInput({ lensType: 'Zoom' });
    const a = legacy(input);
    const b = core(input);
    expect(new Set(b.map(x => x.source_url || x.name))).toEqual(new Set(a.map(x => x.source_url || x.name)));
  });

  it('snapshot: Prime', () => {
    const input = makeInput({ lensType: 'Prime' });
    const b = core(input);
    const keys = b.map(x => x.source_url || x.name).sort();
    expect(keys).toMatchSnapshot();
  });

  it('snapshot: Zoom', () => {
    const input = makeInput({ lensType: 'Zoom' });
    const b = core(input);
    const keys = b.map(x => x.source_url || x.name).sort();
    expect(keys).toMatchSnapshot();
  });
});
