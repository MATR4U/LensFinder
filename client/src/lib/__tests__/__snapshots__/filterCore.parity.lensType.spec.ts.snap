// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parity: lens type gating > snapshot: Prime 1`] = `[]`;

exports[`parity: lens type gating > snapshot: Zoom 1`] = `
[
  "https://example.com/24-70gm",
]
`;
